### PR TITLE
Add PKCE support

### PR DESCRIPTION
**Pull request by whamtet**
Thanks for the great library @weavejester .  Twitter requires a small extension to standard oauth called PKCE ([details](https://oauth.net/2/pkce/)).  I suggest adding a `:pkce?` boolean to the config that auto generates the required parameters when true.  Supplied is a minimal working implementation that satisfies twitter.

If you agree with this feature I will add proper random generation and documentation / unit tests.  I wanted to see whether you agree with the feature first.

### CODE REVIEW

**Comment by weavejester** [comment]
This seems a reasonable feature to add.

**Comment by whamtet** [comment]
Great then I will add it.  The existing implementation already generates a random `state` key.  I can't tell from the documentation whether the `code verifier` referred to (here)[https://www.oauth.com/oauth2-servers/pkce/authorization-request/] is this state, or whether I should generate a second random string.  What do you think @weavejester ?

**Comment by weavejester** [comment]
I'm not a cryptographer, but my understanding is that reusing random numbers for different purposes can be a bad idea. Generating a new number at guarantees there's no connection that might inadvertently cause a security issue.

Also the state is 9 bytes, while the RFC says the code verifier should be at least 32 bytes.

**Comment by whamtet** [comment]
@weavejester any thoughts on this?

File: src/ring/middleware/oauth2.clj [hunk 1]
```diff
@@ -86,7 +116,7 @@
 (defn no-auth-code-handler [_]
   {:status 400, :headers {}, :body "No authorization code"})
 
-(defn- make-redirect-handler [{:keys [id landing-uri] :as profile}]
+(defn make-redirect-handler [{:keys [id landing-uri] :as profile}]
```
Comment (weavejester) [inline src/ring/middleware/oauth2.clj:119]: Is this a typo? I can't see a reason to make this function public.

File: test/ring/middleware/oauth2_test.clj [hunk 2]
```diff
@@ -231,6 +245,29 @@
...
 
+(defn openid-response-with-code-verifier [req]
+  {:status 200
+   :headers {"Content-Type" "application/json"}
+   :body (cheshire/generate-string
+          {:access_token "defdef"
+           :expires_in 3600
+           :refresh_token "ghighi"
+           :id_token "abc.def.ghi"
+           :code_verifier (-> req :body slurp codec/form-decode (get "code_verifier"))})})
```
Comment (weavejester) [inline test/ring/middleware/oauth2_test.clj:256]: Align clauses to be consistent with the style of the rest of the code.

File: test/ring/middleware/oauth2_test.clj [hunk 3]
```diff
@@ -231,6 +245,29 @@
...
+
+(deftest test-openid-response-with-code-verifier
+  (fake/with-fake-routes
+   {"https://example.com/oauth2/access-token" openid-response-with-code-verifier}
+
+   (testing "valid state"
+            (let [verifier (oauth2/code-verifier)
```
Comment (weavejester) [inline test/ring/middleware/oauth2_test.clj:263]: `testing` indentation not correct, should be indented two spaces. Also take care to ensure lines are within 80 characters in this test.

**Review by weavejester — CHANGES_REQUESTED** [review]
Thanks for the update and apologies for taking a while to get back to you. I've reviewed the code and have some suggestions. Also be sure to take a look at the contributing document and adjust the commit messages accordingly. Have you tested this middleware using Twitter's oauth?

**Comment by whamtet** [comment]
Thanks @weavejester I have made all the changes you suggested.

File: README.md [hunk 4]
```diff
@@ -129,6 +129,14 @@ and complete authentication of the user.
 
 [the specification]: https://tools.ietf.org/html/rfc6749#section-2.3.1
 
+### PKCE
+
+Some oauth providers require an additional step called Proof Key for 
+Code Exchange 
+([PKCE](https://www.oauth.com/oauth2-servers/pkce/authorization-request/)).  Ring-OAuth2
+will include a proof key in the workflow when `:pkce? true`is included 
+in the configuration map.
```
Comment (weavejester) [inline README.md:138]: Let's write it as:

```markdown
Some OAuth providers require an additional step called *Proof Key for
Code Exchange* ([PKCE][]). Ring-OAuth2 will include a proof key in the
workflow when `pkce?` is set to `true` in the profile map.

[pkce]: https://www.oauth.com/oauth2-servers/pkce/authorization-request/
```

This is a little neater, is within line limits, and uses the same terminology as the rest of the README.

Comment (whamtet) [reply README.md:138]: oh nice, I didn't realize README had this feature

File: src/ring/middleware/oauth2.clj [hunk 5]
```diff
@@ -16,23 +19,48 @@
...
+       (cond-> {:response_type "code"
+                :client_id     (:client-id profile)
+                :redirect_uri  (redirect-uri profile request)
+                :scope         (scopes profile)
+                :state         state}
+               (:pkce? profile)
+               (assoc
+                :code_challenge (verifier->challenge code-verifier)
+                :code_challenge_method "S256")
+               true codec/form-encode)))
```
Comment (weavejester) [inline src/ring/middleware/oauth2.clj:47]: This is a little hard to read. Instead, let's do:

```clojure
(defn- authorize-params [profile request state verifier]
  (-> {:response_type "code"
       :client_id     (:client-id profile)
       :redirect_uri  (redirect-uri profile request)
       :scope         (scopes profile)
       :state         state}
      (cond-> (:pkce? profile)
        (assoc :code_challenge (verifier->challenge code-verifier)
               :code_challenge_method "S256"))))

(defn- authorize-uri [profile request state verifier]
  (str (:authorize-uri profile)
       (if (.contains ^String (:authorize-uri profile) "?") "&" "?")
       (codec/form-encode (authorize-params profile request state verifier))))
```

**Comment by whamtet** [comment]
> Thanks for the update and apologies for taking a while to get back to you. I've reviewed the code and have some suggestions. Also be sure to take a look at the contributing document and adjust the commit messages accordingly. Have you tested this middleware using Twitter's oauth?

Yes it works nicely with Twitter!

File: src/ring/middleware/oauth2.clj [hunk 6]
```diff
@@ -16,23 +19,49 @@
...
-                           :redirect_uri  (redirect-uri profile request)
-                           :scope         (scopes profile)
-                           :state         state})))
+       (codec/form-encode (authorize-params profile request state verifier))))
 
 (defn- random-state []
-  (-> (random/base64 9) (str/replace "+" "-") (str/replace "/" "_")))
+  (base64url (random/base64 9)))
+
+(defn random-code-verifier []
```
Comment (weavejester) [inline src/ring/middleware/oauth2.clj:53]: This function should be private, right?

Comment (whamtet) [reply src/ring/middleware/oauth2.clj]: yes!

**Review by weavejester — APPROVED** [review]
Thanks, this all looks good. Can you squash your commits down and format the commit message according to [the seven rules of a great git commit message](https://chris.beams.io/posts/git-commit/#seven-rules)?

**Comment by whamtet** [comment]
How about now @weavejester ?

**Comment by weavejester** [comment]
Thanks. Let's change the commit message just a little bit; instead of going into detail, let's cut it short and add a link to the specification:

```
Add PKCE support

Proof Key for Code Exchange (PKCE) is an extension to the Authorization
Code flow to prevent CSRF and authorization code injection attacks. This
commit adds the :pkce? option, disabled by default.

See: https://oauth.net/2/pkce/
```

**Comment by whamtet** [comment]
done!

**Comment by whamtet** [comment]
`random-code-verifier` is needed by the tests so it will be public.  This is my last update to the PR, take it or leave it Mr Pedantic

**Comment by weavejester** [comment]
I admittedly do have high standards when it comes to code I'll be maintaining indefinitely, though not sure I'd classify passing tests as being pedantic. Still, it's your choice not to continue. I'll put finishing this off on my todo list and take a look at it when I next have time.

**Pull request closed without merging**